dimod/sampleset.py

```diff
@@ -1391,6 +1391,90 @@ def slice(self, *slice_args, **kwargs):
         return type(self)(record, self.variables, copy.deepcopy(self.info),
                           self.vartype)
 
+    def append_data_vectors(self, vectors, inplace=True):
+        """Return the :class:`SampleSet` with additional fields in
+        :attr:`SampleSet.record`.
+
+        Args:
+            vectors (dict):
+                Each key is a new field name and each value is a list of 
+                per-sample data to be appended to :attr:`SampleSet.record`. 
+                Per-sample data should be scalar or numpy arrays (lists 
+                and tuples will be converted to numpy arrays).
+
+            inplace (bool, optional, default=True):
+                If True, the instantiated :class:`SampleSet` is updated; 
+                otherwise, a new :class:`SampleSet` is returned.
+
+        Returns:
+            :obj:`.SampleSet`: SampleSet. If `inplace` is True, returns itself.
+
+        Examples:
+            The following example appends a field of lists to :attr:`SampleSet.record`.
+
+            >>> import numpy as np
+            ...
+            >>> sampleset = dimod.SampleSet(np.rec.array([([-1,  1], -1.4, 1), ([-1,  1], -1.4, 1)],
+                                            dtype=[('sample', 'i1', (2,)), ('energy', '<f8'), 
+                                            ('num_occurrences', '<i8')]), [0, 1], {}, 'SPIN')
+            >>> print(sampleset)
+               0  1 energy num_oc.
+            0 -1 +1   -1.4       1
+            1 -1 +1   -1.4       1
+            ['SPIN', 2 rows, 2 samples, 2 variables]
+
+            >>> vector = {'new_col': [[0, 1], [1, 2]]}
+            >>> sampleset.append_data_vectors(vector)
+            >>> print(sampleset)
+               0  1 energy num_oc. new_col
+            0 -1 +1   -1.4       1   [0 1]
+            1 -1 +1   -1.4       1   [1 2]
+            ['SPIN', 2 rows, 2 samples, 2 variables]
+
+            >>> print(sampleset.record.dtype)
+            (numpy.record, [('sample', 'i1', (2,)), 
+                            ('energy', '<f8'), 
+                            ('num_occurrences', '<i8'), 
+                            ('new_col', '<i8', (2,))])
+        
+        """
+        self.resolve()
+
+        record = self._record
+
+        for name, vector in vectors.items():
+            if name in record.dtype.names:
+                raise ValueError("Field '{}' occurs more than once".format(name))
+
```

Comment:
I think this can be simplified with `vector = np.asarray(vector)`, which would give you access to `.shape` and would check that it can be cast to a numpy array etc etc.

Reply:
Hm, I actually do end up doing that on line 1459. The checks above just take other possible errors into account.

Reply:
Right, I guess I find `arr.ndim == 1` more satisfying than `np.isscalar(vector[0])`

Reply:
But I cannot think of an error case where `np.isscalar(vector[0])` would fail off hand so maybe it's fine.

Reply:
On second thought, I agree and have made the change 👌